lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,413 @@
+                else if key == TIMESTAMP {
+                    if !(value.is_timestamp() || value.is_integer()) {
+                        emit_invalid_type(key, "timestamp or integer", value.kind_str())?;
+                    }
+                }
+                // validate integer values
+                else if key == LEVEL {
+                    if !value.is_integer() {
+                        emit_invalid_type(key, "integer", value.kind_str())?;
+                    }
+                }
+                // validate float values
+                else if key == LINE {
+                    if !(value.is_float() || value.is_integer()) {
+                        emit_invalid_type(key, "number", value.kind_str())?;
+                    }
+                } else {
+                    // Additional fields must be prefixed with underscores.
+                    // Prepending the underscore since vector adds fields such as 'source_type'
+                    // which would otherwise throw errors.
+                    if !key.is_empty() && !key.starts_with('_') {
+                        if self.conformed_log.is_none() {
+                            self.conformed_log = Some(log.clone())
+                        }
+                        self.conformed_log
+                            .as_mut()
+                            .unwrap()
+                            .rename_key(key.as_str(), &*format!("_{}", &key));
+                    }
+
+                    // additional fields must be only word chars, dashes and periods.
+                    if !VALID_FIELD_REGEX.is_match(key) {
+                        vector_core::internal_event::emit(SerializeFailedInvalidFieldName {
+                            format_type: FORMAT_TYPE_GELF,
+                            message: INVALID_FIELD_NAME_STR,
+                            field: key,
+                        });
+                        return Err(format!("{}: {}", INVALID_FIELD_NAME_STR, key).into());
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Determine if input log event is in valid GELF format
+    fn validate_event_is_gelf(&mut self, log: &LogEvent) -> vector_core::Result<()> {
+        self.validate_required_fields(log)?;
+        self.validate_field_names_and_values(log)
+    }
+}
+
+impl Default for GelfSerializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Encoder<Event> for GelfSerializer {
+    type Error = vector_core::Error;
+
+    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        let log = event.as_log();
+        let writer = buffer.writer();
+
+        self.validate_event_is_gelf(log)?;
+
+        // if event was conformed to valid GELF, use that instead
+        if let Some(conformed) = &self.conformed_log {
+            serde_json::to_writer(writer, &conformed)?;
+        } else {
+            serde_json::to_writer(writer, &log)?;
+        }
```

Comment:
I think a better pattern here would be to have the function return the transformed event rather than having `validate_event_is_gelf` have a side-effect (setting `self.conformed_log`) that is checked later. We could call it something like `to_gelf_event`. I generally think this makes the control flow easier to follow.

Reply:
I'll look at that approach again. I had that initially but given I have helper methods that can also create the conformed_log , I was placing the conformed_log as a member of the struct to avoid doing the copy unless we really needed to.